#5 : Add easy server ON/OFF button

leo7418 opened the pull request:
https://github.com/orgs/FileShare-Project/projects/2/views/1?pane=issue&itemId=44793914

leo7418 committed:
- New file Application.cpp/hpp:
    -  Display sider and content
    - On sider:
        - Display LAN, Own devices and Contacts sections
        - For each section item, display name and connected/disconnected icon
- On main:
    - Initialize an application
    - Run the application

leo7418 committed:
- On Application:
    - Move init function inside constructor
    - Clean code
- On main:
    - Return 1 only for errors

leo7418 committed:
- Move Sider section from Application class to new class DeviceList
- On DeviceList:
    - Can now set the current device
    - Can now toggle the device connection
- On .gitignore:
    - Add vscode settings folder

leo7418 committed:
- Convert to MVC architecture
- On CMakeLists:
    - Add function add_mvc_executable which add each cpp files for a given folder name
    - Use CMakeLists for DeviceList
- New Makefile:
    - Add functions for clean, build and run the application

f-michaut commented on Makefile:
```diff
@@ -0,0 +1,45 @@
+
+.PHONY: build-run br
+build-run br: build
+	$(call printSection,"Build completé. Lancement du programme...")
```
```suggestion
	$(call printSection,"Build completed. Running Program...")
```

f-michaut commented on include/DeviceList/View.hpp:
```diff
@@ -0,0 +1,45 @@
+            View();
+            ~View();
+
+			typedef std::shared_ptr<View> Ptr;
+			typedef std::shared_ptr<const View> ConstPtr;
+
+			static View::Ptr create() { return std::make_shared<View>(); }
+			static View::Ptr copy(View::ConstPtr widget) { return widget ? std::static_pointer_cast<View>(widget->clone()) : nullptr; }
```
Indentation is off

f-michaut commented on include/DeviceList/View.hpp:
```diff
@@ -0,0 +1,45 @@
+            tgui::Signal& getSignal(tgui::String signalName) override;
+
+            void addDeviceList(const std::string& title, const std::vector<std::string>& devices) { this->createSection(title, devices); }
+			void setCurrentDevice(std::string& device);
+			void setCurrentDeviceConnected(bool connected);
+
+            tgui::SignalTyped<const std::string&> onSelectDevice = { "onSelectDevice" };
+            tgui::Signal onToggleDevice = { "onToggleDevice" };
+
+		protected:
+			tgui::Widget::Ptr clone() const override { return std::make_shared<View>(*this); }
+
+        private:
+			void createCurrentDeviceSection();
+            void createSection(const std::string& title, const std::vector<std::string>& options);
```
same

f-michaut commented on .gitignore:
```diff
@@ -7,3 +7,5 @@ compile_commands.json
 *.gch
 vgcore.*
+
+.vscode/
```
Should end with a line break :p

f-michaut commented on CMakeLists.txt:
```diff
@@ -99,9 +99,23 @@ include_directories(
+    "source/${name}/View.cpp"
+    "source/${name}/Model.cpp"
+    "source/${name}/Controller.cpp"
+    PARENT_SCOPE
```
Do we need `PARENT_SCOPE` ? If I understand correctly, it is only useful for a sub-cmake file to propagate a variable change in a parent cmakefile, but this is the root cmakefile so it has no parent ?

leo7418 replied:
Yes, we need it. When using the set function inside a CMake function (which add a scope), PARENT_SCOPE is necessary to modify the variable in the parent scope.

f-michaut commented on CMakeLists.txt:
```diff
@@ -99,9 +99,23 @@ include_directories(
+set(executables "")
+
+function(add_mvc_executable name)
+  set(executables
```
Instead of `set`, should we use `list(APPEND, ` instead ? [link](https://cmake.org/cmake/help/latest/command/list.html#append)

That way, we could make the function accept a list of `name` or we could call it multiple times with different `name` values if we have more than one MVC later

leo7418 replied:
Yes, I've noticed that as well, but the list command does not support PARENT_SCOPE, which is why we cannot use it to modify a variable in the parent scope

f-michaut commented on CMakeLists.txt:
```diff
@@ -99,9 +99,23 @@ include_directories(
 
+set(executables "")
+
+function(add_mvc_executable name)
```
```suggestion
function(add_mvc_sources name)
```
The function is adding a list of source files, not executables.

leo7418 replied:
I reused the same term as the basic CMake function `add_executable`, which also involves adding source files, although the name is "like that".

Edit: Additionally, I've modified this function to accept a list instead of just a single filename, allowing it to be called once, similar to add_executable, instead of having to call it multiple times for each MVC component that will be developed.

f-michaut replied:
But the `add_executable` creates an executable with a list of sources, so it does something different.

`add_executable` here create the `file-share` executable with the sources `source/main.cpp, source/Application.cpp`. 

The add_mvc_x function does not create an executable, it just adds a list of sources to be included in the final executable

f-michaut commented on include/DeviceList/Model.hpp:
```diff
@@ -0,0 +1,40 @@
+            void setCurrentDevice(const std::string& device) { this->currentDevice = device; }
+            std::string getCurrentDevice() { return this->currentDevice; }
+
+            void toggleCurrentDeviceConnected() { this->currentDeviceConnected = !this->currentDeviceConnected; }
```
Instead of a toggle, I think a `setCurrentDeviceConnected` would be better ?
Toggles can get out of sync, if you set the value then it will always be correct.

leo7418 replied:
In this specific use-case, I cannot implement setCurrentDeviceConnected because the Controller is not supposed to be aware of the data in the Model. I plan to create one later when the Model is connected to the server and thus has "real" data

Edit: I can still make changes using the getter, and the rest of the Model's set function will be completed later

f-michaut commented on source/DeviceList/Controller.cpp:
```diff
@@ -0,0 +1,42 @@
+namespace FileShare::GUI::DeviceList {
+    Controller::Controller() {
+        this->view = View::create();
+        this->view->onSelectDevice.connect([this](const std::string& device) {
+            this->onSelectDevice(device);
+        });
+        this->view->onToggleDevice.connect([this]() {
+            this->onToggleDevice();
+        });
```
```suggestion
        this->view->onSelectDevice.connect([this](const std::string& device) {
            this->HandleDeviceSelected(device);
        });
        this->view->onToggleDevice.connect([this]() {
            this->HandleDeviceToggled();
        });
```
Open to different naming scheme, the goal is to have a different name for the trigger (`onSelectDevice`) and the action (`HandleDeviceSelected`), otherwise it gets confusing.

f-michaut commented on source/main.cpp:
```diff
@@ -2,20 +2,18 @@
 int main()
 {
-    FileShare::GUI::Application* app = nullptr;
+    FileShare::GUI::Application app;
```
💯

leo7418 committed:
- On .gitignore:
    - Remove "/" after .vscode
- On CMakeLists:
    - add_mvc_executable accept now a list of string
- On Makefile:
    - Fix borders of printSection
- On DeviceList model:
    - Transform toggleCurrentDeviceConnected to setCurrentDeviceConnected
- On DeviceList controller:
    - Transform events receiver from on[name] to handle[name]
    - Call setCurrentDeviceConnected by toggling the current value
- On DeviceList view:
    - Fix indentation
- Transform all getters to const methods

f-michaut reviewed (dismissed):
I'm still not convinced about the CMake function name, but otherwise all good

leo7418 committed:
- On CMakeLists:
    - Rename
add_mvc_executable to add_mvc_source
